staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/controller.go

```diff
@@ -0,0 +1,420 @@
+				utilruntime.HandleError(fmt.Errorf("failed to parse key %v: %w", bindingName, err))
+				continue
+			}
+
+			var binding fantasyInternalPolicyBinding
+			if len(bindingNamespace) == 0 {
+				binding, err = c.policyBindingController.
+					Informer().
+					Get(bindingName)
+			} else {
+				binding, err = c.policyBindingController.
+					Informer().
+					Namespaced(bindingNamespace).
+					Get(bindingName)
+			}
+
+			if err != nil {
+				if k8serrors.IsNotFound(err) {
+					// Binding not found from list. Clean it up
+					allBindings.Delete(bindingName)
+					continue
+				} else {
+					// Other, bad error
+					utilruntime.HandleError(fmt.Errorf("cel-admission-controller: failed to look up binding %w", err))
+					continue
+				}
+			}
+
+			if !binding.Matches(a) {
+				continue
+			}
+
+			bindingInfo := c.bindingInfos[bindingName]
+			if bindingInfo.evaluator == nil {
+				// Compile policy definition using binding
+				//!TODO: for each binding should there be an evaluator for each
+				// target GVK?
+				bindingInfo.evaluator, err = definition.Compile(nil)
+				if err != nil {
+					// compilation error. What to do with it?
+					utilruntime.HandleError(fmt.Errorf("failed to compile CEL expression: %w", err))
+					continue
+				}
+				c.bindingInfos[bindingName] = bindingInfo
+			}
+
+			// Fetch params object by indexing
+			paramsGr := definition.ParamSource()
+			paramsNamespace, paramsName := binding.TargetParams()
+			paramInfo, ok := c.paramsCRDControllers[paramsGr]
+			if !ok {
+				// Not watching this binding's params CRD? seems like a bug yo
+				continue
+			}
+
+			var param *unstructured.Unstructured
+			if len(paramsNamespace) == 0 {
+				param, err = paramInfo.controller.Informer().Get(paramsName)
+			} else {
+				param, err = paramInfo.controller.Informer().Namespaced(paramsNamespace).Get(paramsName)
+			}
+
+			if err != nil {
+				if k8serrors.IsNotFound(err) {
+					// Param doesnt exist yet?
+					// Maybe just have to wait a bit.
+					continue
+				}
+
+				// There was a bad internal error
+				utilruntime.HandleError(err)
+				continue
+			}
+
+			decisions := bindingInfo.evaluator(a, param)
+			for _, decision := range decisions {
+				switch decision.PolicyDecisionKind {
+				case Admit:
+					// Do nothing
+				case Warn:
+					// How to surface warning to user? Just log in apiserver logs?
+				case Deny:
+					allDecisions = append(allDecisions, decision)
+				default:
+					// unrecognized decision. ignore
+				}
+			}
+		}
+	}
+
+	if len(allDecisions) > 0 {
+		return errors.New("deny")
+	}
+
+	return nil
+}
+func (c *celAdmissionController) HasSynced() bool {
+	return c.policyBindingController.HasSynced() &&
+		c.policyDefinitionsController.HasSynced()
+}
+
+func (c *celAdmissionController) reconcilePolicyDefinition(namespace, name string, definition fantasyInternalPolicyDefinition) error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	namespacedName := namespace + "/" + name
```

Comment:
This will always be cluster scoped.

If we do add namespace scoped policy definitions, we will have no choice but to add another kind (similar to how `Roles` and `ClusterRoles` work).  I suppose if we do have both then we could funnel the reconciliation of both into common code that handles namespace..

Maybe start simple and just drop namespace?

Reply:
Same applies to bindings, except that I can say with higher confidence that we will eventually have both namespace and cluster scoped bindings in the future.

Reply:
[ref](https://github.com/kubernetes/enhancements/blob/bf5c3c81ea2081d60c1dc7c832faa98479e06209/keps/sig-api-machinery/3488-cel-admission-control/README.md?plain=1#L1042) for only support cluster scope in phase1 and possibly adding NamespacePolicyBinding in phase2.

Reply:
We could keep the namespace here and just make sure it's clear that it might be empty? It works fine the way the code is written, the map keys for cluster scoped are always prefixed by /, but it doesn't hurt anything.

Reply:
I've added a note as well as a link to the ref from cici. Marking this as resolved